### PR TITLE
Update Flask-SQLAlchemy to 2.4.0

### LINKED ISSUE #3404: Update Flask-SQLAlchemy to the newest version
This is derived from issue 3371. Flask-SQLAlchemy update to 2.2 and 2.3 both broke the unit tests. More time and efforts will be added on this for a near-future update.

(update to 2.2 has been fixed for unit testing, will prepare a separate PR for this.)

### PR DESCRIPTION
**Pull request by lbeaufort**
## Summary (required)

Resolves #3404 

- Update Flask-SQLAlchemy to 2.4.0
- [This commit](https://github.com/pallets/flask-sqlalchemy/commit/3fec7535922e687e0518bdbfcd24b6389ab5c3af) changed the default behavior of `create_session` from returning a `SignallingSession` class to returning a session maker factory that takes a `class` argument. This PR changes our custom `RoutingSQLAlchemy` `create_session` logic to return a session maker factory that creates `RoutingSession`-type sessions.
- SQLAlchemy 2.3 no longer supports putting `table_args` in an inherited class and throws the error `sqlalchemy.exc.ArgumentError: Can't place __table_args__ on an inherited class with no table.`. The best solution is putting the `table_args` in the subclasses explicitly. (see https://stackoverflow.com/questions/43832848/cant-define-table-args-on-a-child-class-in-a-single-table-inheritance-setup)

## How to test the changes locally

- Consider setting up another virtual env for testing this, to not mess with your default version.
- `pip install -r requirements.txt`
- Run the API locally, test some endpoints
- I tested this with a manual deploy to `dev`
- Make sure to  uninstall/reinstall Flask-SQLAlchemy when switching between this branch or switching back to your usual virtual environment

## Impacted areas of the application
List general components of the application that this PR will affect:

-  Database queries, tasks

### CODE REVIEW

**Comment by fec-jli** [comment]
Sorry, I just modify one audit.py under models folder and push. cause a lot of commits.

**Comment by fec-jli** [comment]
you did great job to start this hard work. Thanks
a couple of things:
1) I setup new virtual env and clone the latest develop branch. checkout your branch
when run `pip install -r requirements.txt` , got two errors, also see screen shut below:
a)ERROR: elasticsearch 5.5.1 has requirement urllib3<1.22,>=1.21.1, but you'll have urllib3 1.24.3 which is incompatible.
b)ERROR: marshmallow-pagination 0.1.1 has requirement marshmallow-sqlalchemy==0.4.1, but you'll have marshmallow-sqlalchemy 0.15.0 which is incompatible.

2) pytest passed all auto-test. great.

3) endpoints_walk.py
got one failed: '/schedules/schedule_a/{sub_id}/?api_key={api_key}: not working. Status: 400'

4) I also modify audit.py under models folder. test all audit endpoints work well.
don't test other endpoints yet.
Thanks a lot.

File: webservices/common/models/audit.py [hunk 1]
```diff
@@ -41,6 +40,7 @@ def sub_category_list(self):
 
 # endpoint audit-case
 class AuditCaseSubCategory(db.Model):
+    __table_args__ = {"schema": "auditsearch"}
```
Comment (fec-jli) [inline webservices/common/models/audit.py]: Please delete line 43. don't need this line. Thanks

File: webservices/common/models/audit.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ class AuditCaseSubCategory(db.Model):
 
 # endpoint audit-case
 class AuditCaseCategoryRelation(db.Model):
+    __table_args__ = {"schema": "auditsearch"}
```
Comment (fec-jli) [inline webservices/common/models/audit.py]: please remove this line too. Thanks
__table_args__ = {"schema": "auditsearch"}

File: webservices/common/models/audit.py [hunk 3]
```diff
@@ -68,6 +69,7 @@ class AuditCaseCategoryRelation(db.Model):
 
 # endpoint audit-case
 class AuditCase(db.Model):
+    __table_args__ = {"schema": "auditsearch"}
```
Comment (fec-jli) [inline webservices/common/models/audit.py]: please remove this line.Thanks
__table_args__ = {"schema": "auditsearch"}

File: webservices/common/models/audit.py [hunk 4]
```diff
@@ -98,6 +100,7 @@ class AuditCase(db.Model):
 
 # endpoint audit/search/name/candidates
 class AuditCandidateSearch(db.Model):
+    __table_args__ = {"schema": "auditsearch"}
```
Comment (fec-jli) [inline webservices/common/models/audit.py]: please delete line.

File: webservices/common/models/audit.py [hunk 5]
```diff
@@ -108,6 +111,7 @@ class AuditCandidateSearch(db.Model):
 
 # endpoint audit/search/name/committees
 class AuditCommitteeSearch(db.Model):
+    __table_args__ = {"schema": "auditsearch"}
```
Comment (fec-jli) [inline webservices/common/models/audit.py]: please delete this line.

**Comment by pkfec** [comment]
@fec-jli the endpoint `/schedules/schedule_a/{sub_id}/?api_key={api_key}` is failing because of the validation rules that were added to scheduleA endpoint.More info on validation here :PR #3725

@lbeaufort Fixed the  endpoint and pushed the changes to your branch(feature/update-flask-sqlalchemy).hope you don't mind :)

**Comment by lbeaufort** [comment]
@fec-jli I'll make those audit changes, thanks for taking a look!

@pkfec thanks for making the change to the endpoint walk :)

**Comment by fec-jli** [comment]
Great job. I tested all audit endpoints. work well.
and also tested a couple of other endpoints, all good.

just want to know, when you run `run pip install -r requirements.txt`, did you get any error like I got above? 
this may be not related with this ticket. may be my local env difference.
Thanks

**Comment by lbeaufort** [comment]
@fec-jli I don't think these warnings are specific to this branch, but I'm rebuilding `develop` without cache to test.

**Comment by lbeaufort** [comment]
@fec-jli @pkfec can we merge this before it gets too stale?

**Pull request merged**